libass/ass.c

```diff
@@ -1359,6 +1371,48 @@ void ass_flush_events(ASS_Track *track)
     track->parser_priv->read_order_elems = 0;
 }
 
+void ass_configure_gc(ASS_Track *track, int enabled, long long delay)
+{
+    track->parser_priv->gc_enabled = enabled;
+
+    if (0ll <= delay && enabled)
+        track->parser_priv->gc_delay = delay;
+}
+
+void ass_prune_events(ASS_Track *track, long long deadline)
+{
+    long long next_undisplay_ts = track->parser_priv->gc_next_undisplay_ts;
+
+    //fast exit
+    if (deadline < next_undisplay_ts)
+        return;
+
+    bool check_readorder = track->parser_priv->check_readorder;
+    long long undisplay_ts;
+    int start = 0, stop, diff;
+
+    // find again smallest undisplay pts
+    next_undisplay_ts = LLONG_MAX;
+    for (int k = 0; k < track->n_events; ) {
+        while (k < track->n_events && track->events[k].Start + track->events[k].Duration < deadline) {
+            if (check_readorder)
+                clear_read_order_bit(track, track->events[k].ReadOrder);
+            ass_free_event(track, k++);
+        }
+        stop = k;
+        while (k < track->n_events && !((undisplay_ts = track->events[k].Start + track->events[k].Duration) < deadline)) {
+            next_undisplay_ts = FFMIN(next_undisplay_ts, undisplay_ts);
+            ++k;
+        }
+        diff = k - stop;
+        if (stop > start)
+            memmove(&track->events[start], &track->events[stop], sizeof(ASS_Event)*diff);
+        start += diff;
+    }
+    track->parser_priv->gc_next_undisplay_ts = next_undisplay_ts;
+    track->n_events = start;
+}
```

Comment:
Yep this now avoids spurious data moves, thanks. It leaks a bit too much state outside the loop and a little bit of polishing imho improves readability and reduces mutable variables outside the loop to just two:

```C
void ass_prune_events(ASS_Track *track, long long deadline)
{
    if (deadline < track->parser_priv->gc_next_undisplay_ts)
        return;

    const bool check_readorder = track->parser_priv->check_readorder;
    const ASS_Event *events = track->events;
    const int old_n_events = track->n_events;

    long long min_undisplay_remaining = LLONG_MAX;
    int n_kept = 0;

    for (int k = 0; k < old_n_events;) {
        // discardable sequence
        for (; k < old_n_events && events[k].Start + events[k].Duration < deadline; k++) {
            if (check_readorder)
                clear_read_order_bit(track, events[k].ReadOrder);
            ass_free_event(track, k);
        }

        // to-be-kept sequence
        int move_from = k;
        for(; k < old_n_events && events[k].Start + events[k].Duration >= deadline; k++)
            min_undisplay_remaining = FFMIN(min_undisplay_remaining, events[k].Start + events[k].Duration);

        // Relocate kept events
        if (move_from < k) {
            int cnt = k - move_from;
            memmove(events + n_kept, events + move_from, cnt * sizeof(*track->events));
            n_kept += cnt;
        }
    }

    track->n_events = n_kept;
    track->parser_priv->gc_next_undisplay_ts = min_end_remaining;
}
```

I was wondering why `check_readorder` was cached to a local variable eventhough it’s only checked in a single place, but this seems to slightly improve generated instructions inside the loop and we established this should be as fast as possible. So i kept it and added two more shortcuts (though `events` prob helps more with readability and conciseness rather than codegen)*